src/vulkan-renderer/wrapper/device.cpp

```diff
@@ -31,83 +33,334 @@ constexpr float DEFAULT_QUEUE_PRIORITY = 1.0f;
+                                                   const std::vector<const char *> &required_extensions) {
+    if (physical_device_infos.empty()) {
+        throw std::runtime_error("Error: There are no physical devices available!");
+    }
+    std::sort(physical_device_infos.begin(), physical_device_infos.end(), [&](const auto &lhs, const auto &rhs) {
+        return compare_physical_devices(required_features, required_extensions, lhs, rhs);
+    });
+    if (!is_device_suitable(physical_device_infos.front(), required_features, required_extensions, true)) {
+        throw std::runtime_error("Error: Could not determine a suitable physical device!");
     }
+    return physical_device_infos.front().physical_device;
+}
+
+VkPhysicalDevice Device::pick_best_physical_device(const Instance &inst, const VkSurfaceKHR surface,
+                                                   const VkPhysicalDeviceFeatures &required_features,
+                                                   const std::vector<const char *> &required_extensions) {
+    // Put together all data that is required to compare the physical devices
+    const auto physical_devices = vk_tools::get_all_physical_devices(inst.instance());
+    std::vector<DeviceInfo> infos(physical_devices.size());
+    std::transform(physical_devices.begin(), physical_devices.end(), infos.begin(),
+                   [&](const VkPhysicalDevice physical_device) { return build_device_info(physical_device, surface); });
+    return pick_best_physical_device(std::move(infos), required_features, required_extensions);
+}
 
-    m_physical_device = *selected_gpu;
+Device::Device(const Instance &inst, const VkSurfaceKHR surface, const bool prefer_distinct_transfer_queue,
+               const VkPhysicalDevice physical_device, const VkPhysicalDeviceFeatures &required_features,
+               const std::vector<const char *> &required_extensions, const VkPhysicalDeviceFeatures &optional_features)
+    : m_physical_device(physical_device) {
+    assert(physical_device);
 
-    VkPhysicalDeviceProperties graphics_card_properties;
+    if (!is_device_suitable(build_device_info(physical_device, surface), required_features, required_extensions)) {
+        throw std::runtime_error("Error: The chosen physical device {} is not suitable!");
+    }
 
-    // Get the information about that graphics card's properties.
-    vkGetPhysicalDeviceProperties(m_physical_device, &graphics_card_properties);
+    VkPhysicalDeviceProperties physical_device_properties;
+    vkGetPhysicalDeviceProperties(m_physical_device, &physical_device_properties);
 
-    spdlog::trace("Creating device using graphics card: {}", graphics_card_properties.deviceName);
+    spdlog::trace("Creating device using graphics card: {}", physical_device_properties.deviceName);
 
-    m_gpu_name = graphics_card_properties.deviceName;
+    m_gpu_name = physical_device_properties.deviceName;
 
```

Comment:
Gotcha. knew the unnamed namespace is ugly :P
In the unnamed namespace is a function which does exactly this ^^

Would also rename `m_gpu_name` to `m_device_name`.

Reply:
I don't understand?

Reply:
You didn't used `get_physical_device_name`.

Reply:
Oh, right! Will fix this.

Reply:
Done.